omc3/knob_extractor.py

```diff
@@ -0,0 +1,459 @@
+        return KNOBS_TXT_AFS
+
+    raise FileNotFoundError("None of the knobs-definition files are available.")
+
+
+def _load_knobs_dict(file_path: Union[Path, str]) -> KnobsDict:
+    """ Load the knobs-definition file and convert into KnobsDict.
+    Each line in this file should consist of four comma separated entries:
+    madx-name, lsa-name, scaling factor, knob-test value.
+    Alternatively, a TFS-file is also allowed, but needs to have the suffix ``.tfs``.
+
+    Args:
+        file_path (Path): Path to the knobs definition file.
+
+    Returns:
+        Dictionary with LSA names (but with colon instead of /) as
+        keys and KnobEntries (without values) as value.
+    """
+    if Path(file_path).suffix == ".tfs":
+        # just in case someone wants to give tfs files (hidden feature)
+        df = tfs.read_tfs(file_path)
+    else:
+        # parse csv file (the official way)
+        dtypes = {"madx": str, "lsa": str, "scaling": float, "test": float}
+        converters = {'madx': str.strip, 'lsa': str.strip}  # strip whitespaces
+        df = pd.read_csv(file_path, comment="#", names=dtypes.keys(), dtype=dtypes, converters=converters)
+    return _dataframe_to_knobsdict(df)
+
+
+def _dataframe_to_knobsdict(df: pd.DataFrame) -> KnobsDict:
+    """ Converts a DataFrame into the required Dictionary structure.
+
+    Args:
+        df (pd.DataFrame): DataFrame containing at least the columns
+                           'lsa', 'madx', 'scaling' (upper or lowercase)
+
+    Returns:
+        Dictionary with LSA names (but with colon instead of /) as
+        keys and KnobEntries (without values) as value.
+
+    """
+    df.columns = df.columns.astype(str).str.lower()
+    df = df[['lsa', 'madx', 'scaling']].set_index("lsa", drop=False)
+    return {
+        lsa2name(r[0]): KnobEntry(**r[1].to_dict()) for r in df.iterrows()
+    }
+
+
+def _parse_knobs_defintions(knobs_def_input: Optional[Union[Path, str, pd.DataFrame]]) -> KnobsDict:
+    """ Parse the given knob-definitions either from a csv-file or from a DataFrame. """
+    if isinstance(knobs_def_input, pd.DataFrame):
+        return _dataframe_to_knobsdict(knobs_def_input)
+
+    # input points to a file or is None
+    knobs_def_file = _get_knobs_def_file(knobs_def_input)
+    return _load_knobs_dict(knobs_def_file)
+
+
+# Time Tools -------------------------------------------------------------------
+
+def _parse_time(time: str, timedelta: str = None) -> datetime:
+    """ Parse time from given time-input. """
+    t = _parse_time_from_str(time)
+    if timedelta:
+        t = _add_time_delta(t, timedelta)
+    return t
+
+
+def _parse_time_from_str(time_str: str) -> datetime:
+    """ Parse time from given string. """
+    # Now? ---
+    if time_str.lower() == "now":
+        return datetime.now()
+
+    # ISOFormat? ---
+    try:
+        return datetime.fromisoformat(time_str)
+    except (TypeError, ValueError):
+        pass
```

Comment:
Maybe a `LOGGER.debug("Could not parse time string as ISO format")` before we `pass` would be nice.